File: server/src/services/authService.ts
import crypto from "node:crypto";
import { randomUUID } from "node:crypto";
import { pool } from "../db.js";
import type { AdminProfile } from "../types.js";

const TOKEN_SECRET = process.env.ADMIN_TOKEN_SECRET ?? "bingo-admin-secret-change-me";
const TOKEN_TTL_HOURS = Number(process.env.ADMIN_TOKEN_TTL_HOURS ?? "12");
const DEFAULT_ADMIN_USERNAME = process.env.ADMIN_USERNAME ?? "admin";
const DEFAULT_ADMIN_PASSWORD = process.env.ADMIN_PASSWORD ?? "admin12345";
const DEFAULT_ADMIN_DISPLAY_NAME = process.env.ADMIN_DISPLAY_NAME ?? "Floor Manager";

export function hashPassword(password: string, salt?: string): string {
  const saltValue = salt ?? crypto.randomBytes(16).toString("hex");
  const digest = crypto.pbkdf2Sync(password, saltValue, 120000, 32, "sha256").toString("hex");
  return `${saltValue}$${digest}`;
}

export function verifyPassword(password: string, passwordHash: string): boolean {
  const [salt, storedHash] = passwordHash.split("$", 2);
  const candidate = hashPassword(password, salt);
  const expected = `${salt}$${storedHash}`;
  if (candidate.length !== expected.length) return false;
  return crypto.timingSafeEqual(Buffer.from(candidate), Buffer.from(expected));
}

function signPayload(payload: Record<string, unknown>): string {
  const body = Buffer.from(JSON.stringify(payload));
  const signature = crypto.createHmac("sha256", TOKEN_SECRET).update(body).digest("hex");
  return `${body.toString("base64url")}.${signature}`;
}

function decodeToken(token: string): Record<string, string> {
  try {
    const [encodedPayload, signature] = token.split(".", 2);
    const body = Buffer.from(encodedPayload!, "base64url");
    const expected = crypto.createHmac("sha256", TOKEN_SECRET).update(body).digest("hex");
    if (!crypto.timingSafeEqual(Buffer.from(signature!), Buffer.from(expected))) {
      throw new Error("Invalid token signature");
    }
    const payload = JSON.parse(body.toString("utf-8")) as Record<string, string>;
    if (new Date(payload.exp!) < new Date()) {
      throw new Error("Expired");
    }
    return payload;
  } catch {
    const err = new Error("Invalid admin token") as Error & { statusCode: number };
    err.statusCode = 401;
    throw err;
  }
}

function issueAdminToken(admin: { id: string; username: string; display_name: string }): string {
  const expiresAt = new Date(Date.now() + TOKEN_TTL_HOURS * 60 * 60 * 1000);
  return signPayload({
    sub: admin.id,
    username: admin.username,
    display_name: admin.display_name,
    exp: expiresAt.toISOString(),
  });
}

export async function ensureDefaultAdmin(): Promise<void> {
  const existing = await pool.query(
    "SELECT id FROM admin_users WHERE username = $1",
    [DEFAULT_ADMIN_USERNAME],
  );
  if (existing.rowCount === 0) {
    await pool.query(
      `INSERT INTO admin_users (id, username, password_hash, display_name, is_active)
       VALUES ($1, $2, $3, $4, TRUE)`,
      [
        randomUUID(),
        DEFAULT_ADMIN_USERNAME,
        hashPassword(DEFAULT_ADMIN_PASSWORD),
        DEFAULT_ADMIN_DISPLAY_NAME,
      ],
    );
  }
}

export async function authenticateAdmin(
  username: string,
  password: string,
): Promise<{ token: string; admin: AdminProfile }> {
  const result = await pool.query<{
    id: string;
    username: string;
    display_name: string;
    password_hash: string;
    is_active: boolean;
  }>("SELECT * FROM admin_users WHERE username = $1", [username]);

  const admin = result.rows[0];
  if (!admin || !admin.is_active || !verifyPassword(password, admin.password_hash)) {
    const err = new Error("Invalid username or password") as Error & { statusCode: number };
    err.statusCode = 401;
    throw err;
  }

  return {
    token: issueAdminToken(admin),
    admin: {
      id: admin.id,
      username: admin.username,
      display_name: admin.display_name,
    },
  };
}

export async function getAdminFromToken(token: string): Promise<AdminProfile> {
  const payload = decodeToken(token);
  const result = await pool.query<{
    id: string;
    username: string;
    display_name: string;
    is_active: boolean;
  }>("SELECT * FROM admin_users WHERE id = $1", [payload.sub]);

  const admin = result.rows[0];
  if (!admin || !admin.is_active) {
    const err = new Error("Admin account unavailable") as Error & { statusCode: number };
    err.statusCode = 401;
    throw err;
  }

  return {
    id: admin.id,
    username: admin.username,
    display_name: admin.display_name,
  };
}

export async function listAdminUsers() {
  const result = await pool.query(
    `SELECT id, username, display_name, is_active, created_at
     FROM admin_users ORDER BY created_at DESC`,
  );
  return result.rows.map((row) => ({
    id: row.id,
    username: row.username,
    display_name: row.display_name,
    is_active: row.is_active,
    created_at: row.created_at,
  }));
}

export async function createAdminUser(
  username: string,
  password: string,
  displayName: string,
  isActive = true,
) {
  if (password.length < 8) {
    const err = new Error("Password must be at least 8 characters") as Error & { statusCode: number };
    err.statusCode = 400;
    throw err;
  }

  const existing = await pool.query("SELECT id FROM admin_users WHERE username = $1", [username]);
  if (existing.rowCount && existing.rowCount > 0) {
    const err = new Error("Username already exists") as Error & { statusCode: number };
    err.statusCode = 400;
    throw err;
  }

  const inserted = await pool.query(
    `INSERT INTO admin_users (id, username, password_hash, display_name, is_active)
     VALUES ($1, $2, $3, $4, $5)
     RETURNING id, username, display_name, is_active, created_at`,
    [randomUUID(), username, hashPassword(password), displayName || "Admin", isActive],
  );
  return inserted.rows[0]!;
}

export async function updateAdminUser(
  adminId: string,
  displayName: string | null | undefined,
  password: string | null | undefined,
  isActive: boolean | null | undefined,
) {
  const existing = await pool.query("SELECT * FROM admin_users WHERE id = $1", [adminId]);
  const admin = existing.rows[0];
  if (!admin) {
    const err = new Error("Admin user not found") as Error & { statusCode: number };
    err.statusCode = 404;
    throw err;
  }

  if (password != null) {
    if (password.length < 8) {
      const err = new Error("Password must be at least 8 characters") as Error & {
        statusCode: number;
      };
      err.statusCode = 400;
      throw err;
    }
    admin.password_hash = hashPassword(password);
  }
  if (displayName != null && displayName.trim()) {
    admin.display_name = displayName.trim();
  }
  if (isActive != null) {
    admin.is_active = isActive;
  }

  const updated = await pool.query(
    `UPDATE admin_users
     SET display_name = $2, password_hash = $3, is_active = $4
     WHERE id = $1
     RETURNING id, username, display_name, is_active, created_at`,
    [adminId, admin.display_name, admin.password_hash, admin.is_active],
  );
  return updated.rows[0]!;
}

File: app/services/auth_service.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import secrets
from datetime import datetime, timedelta, timezone

from fastapi import HTTPException
from sqlalchemy import select

from app.db import SessionLocal
from app.models.database import AdminUserRecord


TOKEN_SECRET = os.getenv("ADMIN_TOKEN_SECRET", "bingo-admin-secret-change-me")
TOKEN_TTL_HOURS = int(os.getenv("ADMIN_TOKEN_TTL_HOURS", "12"))
DEFAULT_ADMIN_USERNAME = os.getenv("ADMIN_USERNAME", "admin")
DEFAULT_ADMIN_PASSWORD = os.getenv("ADMIN_PASSWORD", "admin12345")
DEFAULT_ADMIN_DISPLAY_NAME = os.getenv("ADMIN_DISPLAY_NAME", "Floor Manager")


def hash_password(password: str, salt: str | None = None) -> str:
    salt_value = salt or secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt_value.encode("utf-8"), 120000)
    return f"{salt_value}${digest.hex()}"


def verify_password(password: str, password_hash: str) -> bool:
    salt, stored_hash = password_hash.split("$", 1)
    candidate = hash_password(password, salt)
    return hmac.compare_digest(candidate, f"{salt}${stored_hash}")


def _sign_payload(payload: dict) -> str:
    body = json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")
    signature = hmac.new(TOKEN_SECRET.encode("utf-8"), body, hashlib.sha256).hexdigest()
    return (
        base64.urlsafe_b64encode(body).decode("utf-8").rstrip("=")
        + "."
        + signature
    )


def _decode_token(token: str) -> dict:
    try:
        encoded_payload, signature = token.split(".", 1)
        padded = encoded_payload + "=" * (-len(encoded_payload) % 4)
        body = base64.urlsafe_b64decode(padded.encode("utf-8"))
        expected_signature = hmac.new(TOKEN_SECRET.encode("utf-8"), body, hashlib.sha256).hexdigest()
        if not hmac.compare_digest(signature, expected_signature):
            raise ValueError("Invalid token signature")
        payload = json.loads(body.decode("utf-8"))
    except Exception as exc:
        raise HTTPException(status_code=401, detail="Invalid admin token") from exc

    expires_at = datetime.fromisoformat(payload["exp"])
    if expires_at < datetime.now(timezone.utc):
        raise HTTPException(status_code=401, detail="Admin session expired")
    return payload


def issue_admin_token(admin: AdminUserRecord) -> str:
    expires_at = datetime.now(timezone.utc) + timedelta(hours=TOKEN_TTL_HOURS)
    payload = {
        "sub": str(admin.id),
        "username": admin.username,
        "display_name": admin.display_name,
        "exp": expires_at.isoformat(),
    }
    return _sign_payload(payload)


def ensure_default_admin() -> None:
    with SessionLocal() as db:
        admin = db.execute(
            select(AdminUserRecord).where(AdminUserRecord.username == DEFAULT_ADMIN_USERNAME)
        ).scalar_one_or_none()
        if admin is None:
            admin = AdminUserRecord(
                username=DEFAULT_ADMIN_USERNAME,
                password_hash=hash_password(DEFAULT_ADMIN_PASSWORD),
                display_name=DEFAULT_ADMIN_DISPLAY_NAME,
                is_active=True,
            )
            db.add(admin)
            db.commit()


def authenticate_admin(username: str, password: str) -> dict:
    with SessionLocal() as db:
        admin = db.execute(
            select(AdminUserRecord).where(AdminUserRecord.username == username)
        ).scalar_one_or_none()

        if admin is None or not admin.is_active or not verify_password(password, admin.password_hash):
            raise HTTPException(status_code=401, detail="Invalid username or password")

        token = issue_admin_token(admin)
        return {
            "token": token,
            "admin": {
                "id": str(admin.id),
                "username": admin.username,
                "display_name": admin.display_name,
            },
        }


def get_admin_from_token(token: str) -> dict:
    payload = _decode_token(token)
    with SessionLocal() as db:
        admin = db.execute(
            select(AdminUserRecord).where(AdminUserRecord.id == payload["sub"])
        ).scalar_one_or_none()
        if admin is None or not admin.is_active:
            raise HTTPException(status_code=401, detail="Admin account unavailable")
        return {
            "id": str(admin.id),
            "username": admin.username,
            "display_name": admin.display_name,
        }


def list_admin_users() -> list[dict]:
    with SessionLocal() as db:
        admins = db.execute(select(AdminUserRecord).order_by(AdminUserRecord.created_at.desc())).scalars().all()
        return [
            {
                "id": str(admin.id),
                "username": admin.username,
                "display_name": admin.display_name,
                "is_active": admin.is_active,
                "created_at": admin.created_at,
            }
            for admin in admins
        ]


def create_admin_user(username: str, password: str, display_name: str, is_active: bool = True) -> dict:
    if len(password) < 8:
        raise HTTPException(status_code=400, detail="Password must be at least 8 characters")
    with SessionLocal() as db:
        existing = db.execute(
            select(AdminUserRecord).where(AdminUserRecord.username == username)
        ).scalar_one_or_none()
        if existing is not None:
            raise HTTPException(status_code=400, detail="Username already exists")

        admin = AdminUserRecord(
            username=username,
            password_hash=hash_password(password),
            display_name=display_name or "Admin",
            is_active=is_active,
        )
        db.add(admin)
        db.commit()
        db.refresh(admin)
        return {
            "id": str(admin.id),
            "username": admin.username,
            "display_name": admin.display_name,
            "is_active": admin.is_active,
            "created_at": admin.created_at,
        }


def update_admin_user(admin_id: str, display_name: str | None, password: str | None, is_active: bool | None) -> dict:
    with SessionLocal() as db:
        admin = db.execute(
            select(AdminUserRecord).where(AdminUserRecord.id == admin_id)
        ).scalar_one_or_none()
        if admin is None:
            raise HTTPException(status_code=404, detail="Admin user not found")

        if display_name is not None and display_name.strip():
            admin.display_name = display_name.strip()
        if password is not None:
            if len(password) < 8:
                raise HTTPException(status_code=400, detail="Password must be at least 8 characters")
            admin.password_hash = hash_password(password)
        if is_active is not None:
            admin.is_active = is_active

        db.commit()
        db.refresh(admin)
        return {
            "id": str(admin.id),
            "username": admin.username,
            "display_name": admin.display_name,
            "is_active": admin.is_active,
            "created_at": admin.created_at,
        }
